data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/ScanObjectWorker.java

```diff
@@ -36,118 +37,126 @@ public class ScanObjectWorker implements Runnable{
-          return s3Client.listObjectsV2(listObjectsV2Request.fetchOwner(true).build()).contents().stream().map(S3Object::key)
-                .filter(path -> !path.endsWith("/"))
-                .filter(includeKeyPath -> excludeKeyPaths.stream()
-                        .noneMatch(excludeItem -> includeKeyPath.endsWith(excludeItem))).collect(Collectors.toList());
-    }
+    private void startProcessingObject(final int waitTimeMillis) {
+        final Optional<SourcePartition<S3SourceProgressState>> objectToProcess = sourceCoordinator.getNextPartition(this::provideFilteredBucketKeyPartitions);
 
-    private void processS3ObjectKeys(final S3ObjectReference s3ObjectReference,
-                                     final S3ObjectHandler s3ObjectHandler,
-                                     final ScanOptions scanOptions){
-        final S3ObjectDetails s3ObjDetails = getS3ObjectDetails(s3ObjectReference);
-        final boolean isKeyMatchedBetweenTimeRange = isKeyMatchedBetweenTimeRange(s3ObjDetails.getS3ObjectLastModifiedTimestamp(),
-                scanOptions.getUseStartDateTime(),
-                scanOptions.getUseEndDateTime());
-        if(isKeyMatchedBetweenTimeRange && (isKeyNotProcessedByS3Scan(s3ObjDetails))){
-            updateKeyProcessedByS3Scan(s3ObjDetails);
-            try{
-                s3ObjectHandler.parseS3Object(s3ObjectReference,null);
-            }catch (IOException ex){
-                deleteKeyProcessedByS3Scan(s3ObjDetails);
-                LOG.error("Error while process the parseS3Object. ",ex);
+        if (objectToProcess.isEmpty()) {
+            try {
+                Thread.sleep(waitTimeMillis);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
             }
+            return;
+        }
+
+        final String bucket = objectToProcess.get().getPartitionKey().split("\\|")[0];
+        final String objectKey = objectToProcess.get().getPartitionKey().split("\\|")[1];
+
+        try {
+            processS3Object(S3ObjectReference.bucketAndKey(bucket, objectKey).build());
+            sourceCoordinator.completePartition(objectToProcess.get().getPartitionKey());
+        } catch (final PartitionNotOwnedException | PartitionNotFoundException | PartitionUpdateException e) {
+            LOG.warn("S3 scan object worker received an exception from the source coordinator. There is a potential for duplicate data from {}, giving up partition and getting next partition: {}", objectKey, e.getMessage());
+            sourceCoordinator.giveUpPartitions();
         }
     }
-    /**
-     * Method will identify already processed key.
-     * @return boolean
-     */
-    private boolean isKeyNotProcessedByS3Scan(final S3ObjectDetails s3ObjectDetails) {
-        return stateSaveMap.get(s3ObjectDetails.getBucket()+s3ObjectDetails.getKey()) == null;
+
+    private List<PartitionIdentifier> provideFilteredBucketKeyPartitions() {
```

Comment:
This seems like a very different concern that reading the objects. I think we can move this into its own class and test in an isolated manner.

Reply:
I think that's a little overkill but sure